src/realm/util/file.cpp

```diff
@@ -1064,64 +1064,47 @@ bool File::is_same_file(const File& f) const
-        DWORD vol_serial_num = file_info.dwVolumeSerialNumber;
-        DWORD file_ndx_high = file_info.nFileIndexHigh;
-        DWORD file_ndx_low = file_info.nFileIndexLow;
-        if (GetFileInformationByHandle(f.m_handle, &file_info)) {
-            return vol_serial_num == file_info.dwVolumeSerialNumber && file_ndx_high == file_info.nFileIndexHigh &&
-                   file_ndx_low == file_info.nFileIndexLow;
-        }
+    // First try the Windows Server 2012 version
+    FILE_ID_INFO file_id_info;
+    if (GetFileInformationByHandleEx(m_handle, FileIdInfo, &file_id_info, sizeof file_id_info)) {
+        UniqueID id{file_id_info.VolumeSerialNumber};
+        memcpy(&id.inode[0], file_id_info.FileId.Identifier, sizeof(file_id_info.FileId.Identifier));
+        return id;
     }
 
     DWORD err = GetLastError(); // Eliminate any risk of clobbering
-    std::string msg = get_last_error_msg("GetFileInformationByHandleEx() failed: ", err);
-    throw std::runtime_error(msg);
-
-#else // POSIX version
-
-    struct stat statbuf;
-    if (::fstat(m_fd, &statbuf) == 0) {
-        dev_t device_id = statbuf.st_dev;
-        ino_t inode_num = statbuf.st_ino;
-        if (::fstat(f.m_fd, &statbuf) == 0)
-            return device_id == statbuf.st_dev && inode_num == statbuf.st_ino;
+    if (err != ERROR_INVALID_PARAMETER) {
+        std::string msg = get_last_error_msg("GetFileInformationByHandleEx() failed: ", err);
+        throw std::runtime_error(msg);
     }
-    int err = errno; // Eliminate any risk of clobbering
-    std::string msg = get_errno_msg("fstat() failed: ", err);
-    throw std::runtime_error(msg);
 
-#endif
-
-    /*
-    FIXME: Here is how to do it on Windows Server 2012 and onwards. This new
-    solution correctly handles file identification on ReFS.
-
-    FILE_ID_INFO file_id_info;
-    if (GetFileInformationByHandleEx(m_handle, FileIdInfo, &file_id_info, sizeof file_id_info)) {
-        ULONGLONG vol_serial_num = file_id_info.VolumeSerialNumber;
-        EXT_FILE_ID_128 file_id     = file_id_info.FileId;
-        if (GetFileInformationByHandleEx(f.m_handle, FileIdInfo, &file_id_info,
-                                         sizeof file_id_info)) {
-            return vol_serial_num == file_id_info.VolumeSerialNumber &&
-                file_id == file_id_info.FileId;
-        }
+    // Fall back to the older function on other versions of Windows
+    BY_HANDLE_FILE_INFORMATION file_info;
+    if (!GetFileInformationByHandle(m_handle, &file_info)) {
+        DWORD err = GetLastError(); // Eliminate any risk of clobbering
+        std::string msg = get_last_error_msg("GetFileInformationByHandle() failed: ", err);
+        throw std::runtime_error(msg);
     }
-    */
-}
 
-File::UniqueID File::get_unique_id() const
-{
-    REALM_ASSERT_RELEASE(is_attached());
-#ifdef _WIN32 // Windows version
-    throw std::runtime_error("Not yet supported");
+    UniqueID id{file_info.dwVolumeSerialNumber};
+    memcpy(&id.inode[0],
```

Comment:
Please use memcpy_s or make an assert that the destination is large enough. Same below.